modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/dump/writer/AbstractDumpWriter.java

```diff
@@ -0,0 +1,232 @@
+import com.enonic.xp.blob.NodeVersionKey;
+import com.enonic.xp.blob.Segment;
+import com.enonic.xp.branch.Branch;
+import com.enonic.xp.repo.impl.dump.DumpConstants;
+import com.enonic.xp.repo.impl.dump.FilePaths;
+import com.enonic.xp.repo.impl.dump.PathRef;
+import com.enonic.xp.repo.impl.dump.RepoDumpException;
+import com.enonic.xp.repo.impl.dump.blobstore.DumpBlobStore;
+import com.enonic.xp.repo.impl.dump.model.BranchDumpEntry;
+import com.enonic.xp.repo.impl.dump.model.CommitDumpEntry;
+import com.enonic.xp.repo.impl.dump.model.DumpMeta;
+import com.enonic.xp.repo.impl.dump.model.VersionsDumpEntry;
+import com.enonic.xp.repo.impl.dump.serializer.DumpSerializer;
+import com.enonic.xp.repo.impl.dump.serializer.json.DumpMetaJsonSerializer;
+import com.enonic.xp.repo.impl.dump.serializer.json.JsonDumpSerializer;
+import com.enonic.xp.repo.impl.node.NodeConstants;
+import com.enonic.xp.repository.RepositoryId;
+import com.enonic.xp.repository.RepositorySegmentUtils;
+
+public abstract class AbstractDumpWriter
+    implements DumpWriter
+{
+    private final static Logger LOG = LoggerFactory.getLogger( FileDumpWriter.class );
+
+    private final DumpBlobStore dumpBlobStore;
+
+    private final BlobStore blobStore;
+
+    private final DumpSerializer serializer;
+
+    protected final FilePaths filePaths;
+
+    protected TarArchiveOutputStream tarOutputStream;
+
+    protected AbstractDumpWriter( final BlobStore blobStore, FilePaths filePaths, DumpBlobStore dumpBlobStore )
+    {
+        this.dumpBlobStore = dumpBlobStore;
+        this.serializer = new JsonDumpSerializer();
+        this.blobStore = blobStore;
+        this.filePaths = filePaths;
+    }
+
+    @Override
+    public void writeDumpMetaData( final DumpMeta dumpMeta )
+    {
+        final PathRef dumpMetaFile = filePaths.metaDataFile();
+
+        try (final OutputStream outputStream = openMetaFileStream( dumpMetaFile ))
+        {
+            outputStream.write( new DumpMetaJsonSerializer().serialize( dumpMeta ) );
+        }
+        catch ( IOException e )
+        {
+            throw new RepoDumpException( "Cannot write dump-meta file", e );
+        }
+    }
+
+    @Override
+    public void openBranchMeta( final RepositoryId repositoryId, final Branch branch )
+    {
+        final PathRef branchMetaPath = filePaths.branchMetaPath( repositoryId, branch );
+        openTarStream( branchMetaPath );
+    }
+
+    @Override
+    public void openVersionsMeta( final RepositoryId repositoryId )
+    {
+        final PathRef versionMetaPath = filePaths.versionMetaPath( repositoryId );
+        openTarStream( versionMetaPath );
+    }
+
+    @Override
+    public void openCommitsMeta( final RepositoryId repositoryId )
+    {
+        final PathRef commitMetaPath = filePaths.commitMetaPath( repositoryId );
+        openTarStream( commitMetaPath );
+    }
+
+    @Override
+    public void closeMeta()
+    {
+        try
+        {
+            this.tarOutputStream.close();
```

Comment:
what about to check on NULL before closing?

Reply:
NPE is better because it will indicate a bug in the code.